algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+    simp at h2,
+    intro x,
+    specialize h2 (is_ideal.to_coset S x),
+    simp [is_ideal.to_coset] at h2,
+    rw is_ideal.coset_zero_iff_mem at h2,
+    split,
+    intro hx,
+    rw ← h2,
+    exact ⟨x, hx, @setoid.refl _ (is_ideal.setoid S) x⟩,
+    exact λ hx, hs hx,
+  right,
+    apply set.ext,
+    intro x,
+    rw set.set_eq_def at h2,
+    simp at *,
+    specialize h2 (is_ideal.to_coset S x),
+    cases h2 with y hy,
+    cases hy with hy1 hy2,
+    rw ← sub_eq_zero at hy2,
+    simp at hy2,
+    calc x = y - (y + -x) : by norm_num
+       ... ∈ T : is_ideal.sub_mem hy1 (hs $ is_ideal.mem_of_coset_zero _ _ hy2)
+end
+
+variable α
+
+def quotient_zero_isomorphism : α/(zero_ideal α) ≅ α :=
+{ f   := @quotient.lift α α (is_ideal.setoid $ zero_ideal α) id
+    (λ x y hxy, by simpa [add_neg_eq_zero] using hxy),
+  g   := is_ideal.to_coset (zero_ideal α),
+  hf  := by refine
+    { map_add := λ x y, _,
+      map_mul := λ x y, _,
+      map_one := rfl };
+    { cases is_ideal.coset_rep x with m hm,
+      cases is_ideal.coset_rep y with n hn,
+      rw [← hm, ← hn],
+      simp },
+  hg  := is_ideal.to_quotient (zero_ideal α),
+  hfg := λ x, rfl,
+  hgf := λ x, let ⟨m, hm⟩ := is_ideal.coset_rep x in
+    by rw ← hm; refl }
+
+def zero_prime_iff_integral_domain : is_prime_ideal (zero_ideal α) ↔ is_integral_domain α :=
+⟨λ ⟨hne, hmul⟩, ⟨λ hzo, hne $ set.ext $ by simpa using (zero_ring_of_zero_eq_one hzo).eq_zero,
+   λ x y, by simpa using hmul⟩,
+ λ ⟨hne, hmul⟩, ⟨λ hzu, hne $ by simp [set.set_eq_def] at hzu; rw hzu 1, by simpa⟩⟩
+
+instance is_prime_ideal.hom_preimage {α : Type u} {β : Type v} [comm_ring α] [comm_ring β]
+(f : α → β) [is_hom f] (S : set β) [is_ideal β S] [is_prime_ideal S] :
+  @is_prime_ideal α _ ((f)⁻¹' S) (is_ideal.hom_preimage f S) :=
+let ⟨hnu, hmul⟩ := _inst_7 in
+⟨λ h, have (1:α) ∈ f ⁻¹' S, by rw h; trivial,
+   hnu $ is_ideal.univ_of_one_mem S $ by simpa [is_hom.map_one f] using this,
+ λ x y, by simpa [is_hom.map_mul f] using hmul⟩
+
+theorem is_field.to_is_integral_domain : is_field α → is_integral_domain α :=
+begin
+  intro h,
+  cases h,
+  constructor,
+  exact h_zero_ne_one,
+  intros x y hxy,
+  rw @@or_iff_not_and_not (classical.prop_decidable _) (classical.prop_decidable _),
+  intro h,
+  cases h with hx hy,
+  apply hx,
+  specialize h_h hy,
+  cases h_h with z hz,
+  exact calc
+      x = x * 1 : eq.symm (mul_one x)
+    ... = x * (y * z) : congr_arg _ hz.symm
+    ... = (x * y) * z : eq.symm (mul_assoc x y z)
+    ... = 0 * z : congr_arg (λ m, m * z) hxy
+    ... = 0 : zero_mul z
+end
+
+variable {α}
+
+theorem is_maximal_ideal.to_is_prime_ideal : is_maximal_ideal S → is_prime_ideal S :=
+by rw [maximal_iff_quotient_field, prime_iff_quotient_integral_domain]; exact is_field.to_is_integral_domain (α/S)
+
+theorem not_unit_of_mem_maximal_ideal : is_maximal_ideal S → S ⊆ nonunits α :=
+λ hs x hx hxy, @@is_maximal_ideal.ne_univ_ideal _ S _ hs $ is_ideal.eq_univ_of_contains_unit S ⟨x, hx, hxy⟩
+
```

Comment:
The correct theorem is that a proper ideal is contained in the non-units; your theorem is a trivial corollary of this.